DM-45775: Enable datastore cache for client/server butler

timj opened the pull request:
## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

timj committed:
Enable datastore cache for client/server butler

dhirving commented on python/lsst/daf/butler/remote_butler/_remote_butler.py:
```diff
@@ -171,6 +178,25 @@ def dimensions(self) -> DimensionUniverse:
+                    datastore_config["cached"], universe=self.dimensions
+                )
+            else:
+                self._datastore_cache_manager = DatastoreDisabledCacheManager("", universe=self.dimensions)
```
What are the mechanics by which the tests are always going through the `"cached" in datastore_config` branch and not going through this `else`... where is the `cached` key coming from?  I don't see it in `configs/datastore.yaml` or `DatastoreConfig`.

timj replied:
It comes from `datastores/fileDatastore.yaml`, hence the comment about the default datastore config has to be a FileDatastore config.

dhirving replied:
Oh that makes sense.  I didn't understand that the individual configs load their defaults starting from the top-level default butler config.

dhirving commented on python/lsst/daf/butler/remote_butler/_remote_butler.py:
```diff
@@ -171,6 +178,25 @@ def dimensions(self) -> DimensionUniverse:
             return cache.dimensions
 
+    @property
+    def datastore_cache_manager(self) -> AbstractDatastoreCacheManager:
```
This should probably be private since there isn't a good reason for a user to touch this and there isn't an equivalent interface in `DirectButler`.

timj replied:
Yes. I had it private to start with but then learned that from `__init__` you can't ask for the dimension universe. I'll come up with two internal names.

dhirving commented on python/lsst/daf/butler/remote_butler/_remote_butler.py:
```diff
@@ -171,6 +178,25 @@ def dimensions(self) -> DimensionUniverse:
+            datastore_config = DatastoreConfig()
+            self._datastore_cache_manager: AbstractDatastoreCacheManager
+            if "cached" in datastore_config:
+                self._datastore_cache_manager = DatastoreCacheManager(
```
Won't this cause out-of-disk kills in services using RemoteButler?  Because this creates a new cache manager for every RemoteButler instance, by default this will create a new cache directory for every request and never clean up.

Even if it chose a single cache directory, I think it's undesirable for services to have caching enabled by default since they usually have very little `/tmp` available, generally won't be looking at the same file twice, and have many Butler instances operating simultaneously in different threads.

One idea I had (in [DM-42066](https://rubinobs.atlassian.net/issues/DM-42066)) was to have instantiations via `Butler()` enable the cache by default, and instantiatations via `LabeledButlerFactory`disable it by default... since the difference in intended use cases between these constructors should map pretty well to when it's a good idea to have the cache or not.

timj replied:
I've been assuming that services will set the relevant environment variables like we are going to be doing for user's notebooks. Turning off caching through environment variables can be done. We have the option of defining a separate cache manager config model that we configure to be disabled (ie keep 0 datasets) which would allow environment variables to automatically turn it on if needed. If we always use the disabled cache manager for the factory we make it impossible for anyone to use the environment to tune.

dhirving replied:
I didn't see an environment variable that could be used to disable the cache... which one is it?

It doesn't need to be force-disabled when you instantiate via the factory -- just default disabled if there is no explicit configuration provided.

I don't think it's good to require all consumers to remember to set an environment variable in order for their stuff not to crash out... it would be better if it did the safe thing by default.  If we do want to require everyone to set an envvar then it would be good for this to be accompanied by a Phalanx PR setting it everywhere it is required.

timj replied:
Currently asking for 0 datasets will effectively turn it off but I do need to add a "disabled" mode that can be used to make it explicitly disabled regardless of other settings. 

Sounds like we want to make the RemoteButler take the cache manager as a parameter and in the factory I can create it with a default disabled config but using DatastoreCacheManager so that environment variables can still be configured and respected. If it's not passed in by the constructor the code we have here can trigger so that client users can benefit from the cache.

dhirving replied:
Yeah that sounds OK to me, though I do wonder if we should wait to add the option to enable it for factory-instantiated RemoteButlers until someone asks for it.  I can't think of a situation where this cache in its current form would be useful for a service -- I suspect we might end up writing a different implementation anyway, if someone needs it.

dhirving commented on python/lsst/daf/butler/remote_butler/_remote_butler.py:
```diff
@@ -171,6 +178,25 @@ def dimensions(self) -> DimensionUniverse:
+        # Read the Datastore default config (which is a FileDatastore)
+        # and obtain the default cache manager configuration. Overrides will
+        # come from environment variables. This will not work properly if the
+        # defaults for DatastoreConfig no longer include the cache.
```
Given this, it might be a good idea to add a test showing that the cache is getting used.

timj replied:
I don't think FileDatastore tests that the cache has really been populated either. All the tests of the cache manager are tests of the cache manager.  Should be possible to do a butler.get and ask the remote butler for its cache manager to work out the directory it's using and see if there is a file in there (or use an environment variable to override it).

dhirving replied:
Maybe just add a mock to check that the expected cache manager is getting passed down to `get_dataset_as_python_object`?

timj committed:
Add ability to configure standard datastore cache to be disabled

This allows code to use the normal cache with disabled configruation
but allow external environment variables to turn it on. This is
not possible if the code is using an explicit disabled cache
manager class.

timj committed:
Add new method to create by-default disabled cache manager

That can be enabled via the environment.

timj committed:
Use internal property and internal method for cache_manager

timj committed:
Allow RemoteButler to use disabled datastore cache

timj committed:
Allow remote butler factory to set datastore cache status

timj committed:
Increase test coverage of cache_manager

timj committed:
Change log level of test debug message

timj committed:
Ensure that the env vars are cleared when running cache tests

dhirving reviewed (approved):
Looks OK to me

timj committed:
Add news fragment
